File: rust-2022/src/day02.rs
type I = u8;

pub fn read(inp: &[u8]) -> Vec<[I; 2]> {
    let mut v = Vec::with_capacity(4096);
    for line in inp.split(|&c| c == b'\n') {
        v.push([line[0] - b'A', line[2] - b'X']);
    }
    v
}

pub fn part1(inp: &[[I; 2]]) -> u16 {
    inp.iter().fold(0u16, |acc, &[you, me]| {
        acc + ((me + 1 + 3 * (((me + 4) - you) % 3)) as u16)
    })
}

pub fn part2(inp: &[[I; 2]]) -> u16 {
    inp.iter().fold(0u16, |acc, &[you, result]| {
        acc + (((you + result + 2) % 3 + 1 + 3 * result) as u16)
    })
}

pub fn run() -> (String, String) {
    let inp = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day02.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}

File: 2022/day02.py
import time

start = time.time()
with open("../inputs/2022/day02.input") as f:
    inp = [tuple(ord(c) - ord(b) for c, b in zip(line.split(), "AX")) for line in f]


# Rock-paper-scissors
def part1(inp: list[tuple[int, int]]) -> int:
    # rock == 0, paper == 1, scissors == 2
    # score is 0 for loss, 3 for draw, 6 for win
    # "shape" score is shape + 1
    return sum(me + 1 + 3 * ((me - you + 1) % 3) for you, me in inp)


def part2(inp: list[tuple[int, int]]) -> int:
    # back out desired shape from result
    # scoring as for part 1
    return sum((you + result + 2) % 3 + 1 + 3 * result for you, result in inp)


print(
    f"Day 02     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)
